proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,692 @@
+they are not required by the client's data model.  (For example, in the
+"sharding endpoints for scalability" use-case, different clients may see
+different variants of the EDS resource, but once a given client has the
+right variant, it will be unique on that client, which means that the
+CDS resource does not need to refer to different EDS resource names on
+a different client.)
+
+It should be noted that caching xDS proxies, unlike "leaf" clients, will
+need to track multiple variants of each resource, since a given caching
+proxy may be serving clients that need different variants of a given
+resource.  However, since caching xDS proxies deal with resources only
+at the transport protocol layer, the resource graph layer is
+essentially irrelevant in that case.
+
+### Dynamic Parameters
+
+With the above property in mind, this document proposes the following
+data structures:
+- **Dynamic parameters**, which are a set of key/value pairs sent by the
+  client when subscribing to a resource.
+- **Dynamic parameter constraints**, which are a set of criteria that
+  can be used to determine whether a set of dynamic parameters matches
+  the constraints.  These constraints are part of the cache key for an
+  xDS resource (in addition to the resource name itself) on xDS servers,
+  xDS clients, and xDS caching proxies.  This provides a mechanism to
+  represent multiple variants of a given resource in a cacheable way.
+
+Both of these data structures are used in the xDS transport protocol,
+but they are not part of the resource name and therefore do not appear as
+part of the resource graph.
+
+When a client subscribes to a resource, it specifies a set of dynamic
+parameters.  In response, the server will send a resource whose dynamic
+parameter constraints match the dynamic parameters in the subscription
+request.  The client will use the dynamic parameter constraints on the
+returned resource to determine which of its subscriptions the resource is
+associated with.
+
+#### Constraints Representation
+
+Dynamic parameter constraints will be represented in protobuf form as follows:
+
+```proto
+message DynamicParameterConstraints {
+  // A list of constraints that may be combined with AND or OR semantics.
+  message ConstraintList {
+    // A constraint for a given key.
+    message Constraint {
+      message Exists {}
+      // The key to match against.
+      string key = 1;
+      // How to match.
+      oneof constraint_type {
+        // Matches this exact value.
+        string value = 2;
+        // Key is present (matches any value except for the key being absent).
+        Exists exists = 3;
+      }
+      // If set to true, the match is inverted -- i.e., the key must NOT
+      // match the specified value.
+      bool invert = 4;
+    }
+
+    enum MatchType {
+      // Default value.
+      MATCH_TYPE_UNSPECIFIED = 0;
+      // Logical AND of constraints.
+      MATCH_TYPE_AND = 1;
```

Comment:
Sorry, forgot to respond to this one.

This is a good point.  I'd like some input from @fuqianggao on this, since this was based on a similar internal proto that uses this same format.  If it's really not useful, I'm happy to get rid of it, but I'd like to first verify that this wasn't added for some particular reason internally.

Reply:
I think it is for better UX.

Reply:
Can you say more about why you think it's better?  It's not obvious to me what the benefit is.

Reply:
Actually, we can't get rid of this. Because there are two levels. This MatchType applies to Constraint. While DynamicParameterConstraints.constraints is repeated ConstaintList. Maybe we should name it to constraint_lists to reduce some confusion?

Reply:
Okay, great, if we agree that we can remove this, then we can simplify a bit.  Thanks!

Before I make any changes, I'll wait for more feedback from @htuch and @mattklein123 on the thread below about how extensible we want to make this, given the implications for cache performance and difficulty of making changes later.  I think that once we have consensus on where to land on this spectrum, it should be easier to spell out the details.

Reply:
@fuqianggao Please take a look at the latest changes and verify that they should be implementable for TD.